transmission.go

```diff
@@ -126,78 +127,207 @@ type batch struct {
+	buf := bytes.Buffer{}
+	buf.WriteByte('{')
+
+	keys := make([]string, 0, len(b.events))
+	for datasetName := range b.events {
+		keys = append(keys, datasetName)
+	}
+	sort.Strings(keys) // parity with encoding/json's mapEncoder
+
+	var mapCt int
+	var arrCt int
+	var totalEncoded int
+	for _, datasetName := range keys {
+		kBytes, err := json.Marshal(datasetName) // lean on encoding/json's stringEncoder
+		if err != nil {
+			// return err for all of the dataset's events
+			for _, ev := range b.events[datasetName] {
+				b.enqueueResponse(Response{
+					Err:      err,
+					Metadata: ev.Metadata,
+				})
+			}
+			continue
+		}
+		if mapCt > 0 {
+			buf.WriteByte(',')
+		}
+		mapCt++
+		buf.Write(kBytes)
+		buf.WriteByte(':')
+		buf.WriteByte('[')
 
-	for _, e := range b.events {
-		b.sendRequest(e)
+		arrCt = 0
+		for _, ev := range b.events[datasetName] {
+			eBytes, err := json.Marshal(ev)
+			if err != nil {
+				b.enqueueResponse(Response{
+					Err:      err,
+					Metadata: ev.Metadata,
+				})
+				continue
+			}
+			if arrCt > 0 {
+				buf.WriteByte(',')
+			}
+			arrCt++
+			totalEncoded++
+			buf.Write(eBytes)
+		}
+		buf.WriteByte(']')
+	}
+	if totalEncoded == 0 {
+		// If the datasetName was written but no Events were, make sure to return an
+		// explicitly empty object
+		return []byte("{}"), nil
 	}
+	buf.WriteByte('}')
+	return buf.Bytes(), nil
 }
 
-// sendRequest sends an individual request to Honeycomb and returns
-func (b *batch) sendRequest(e *Event) {
+func (b *batch) enqueueResponse(resp Response) {
+	if b.blockOnResponses {
+		responses <- resp
+	} else {
+		select {
+		case responses <- resp:
+		default: // drop on the floor (and maybe notify tests)
+			if b.testBlocker != nil {
+				b.testBlocker.Done()
+			}
+		}
+	}
+}
+
+func (b *batch) Fire(notifier muster.Notifier) {
+	defer notifier.Done()
+
 	start := time.Now().UTC()
 	if b.testNower != nil {
 		start = b.testNower.Now()
 	}
-	timestamp := e.Timestamp
-	blob, err := json.Marshal(e.data)
-	if err != nil {
-		// TODO add logging or something to raise this error
-		sd.Increment("json_marshal_errors")
+
+	blob, _ := json.Marshal(b) // always returns nil err
+
+	// If json.Marshal returns the shortest valid JSON object, we don't have
+	// anything worth sending to Honeycomb (we may have failed to enqueue any
+	// events, for any dataset). Don't bother enqueueing a Response;
+	// batch.MarshalJSON likely already did that for us.
+	if len(blob) <= 2 {
 		return
 	}
 
+	// sigh. dislike
 	userAgent := fmt.Sprintf("libhoney-go/%s", version)
 	if UserAgentAddition != "" {
 		userAgent = fmt.Sprintf("%s %s", userAgent, strings.TrimSpace(UserAgentAddition))
 	}
 
-	url := fmt.Sprintf("%s/1/events/%s", e.APIHost, e.Dataset)
-	req, err := http.NewRequest("POST", url, bytes.NewBuffer(blob))
-	req.Header.Set("User-Agent", userAgent)
+	// FML. technically we have nothing that constrains events to a single APIHost
+	// or a single WriteKey per batch. ugggg ben suggestions?
```

Comment:
i think it does. i mean, we could introduce even more hierarchy (instead of a map of datasetNames -> events, we could have... what? a map of apihost/writekey -> map of datasetNames -> events)?

gross. but correct! but gross. punting on this for now in case @maplebed has opinions